### PR TITLE
Task: Update docs

### LINKED ISSUE #233: Revamped and more streamlined docs
**Is your feature request related to a problem? Please describe.**
Our documentation
- feels disorganised/disjointed
- lacks a landing page
- lacks the "one-directional" explanation of biotope-element

**Describe the solution you'd like**
A revamped documentation!

### PR DESCRIPTION
**Pull request by tiagomapmarques**
Fixes #233

Docs link: https://deploy-preview-236--biotope-element.netlify.com/docs/next/

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -3,11 +3,22 @@
 [![Build Status](https://travis-ci.org/biotope/biotope-element.svg?branch=master)](https://travis-ci.org/biotope/biotope-element)
 [![codecov](https://codecov.io/gh/biotope/biotope-element/branch/master/graph/badge.svg)](https://codecov.io/gh/biotope/biotope-element)
 
+`biotope-element` is a library to create Web Components in a simple way, taking advantage of simple
+yet powerful concepts from other frontend libraries to empower you, the developer, and your clients'
```
Comment (GomesMiguel) [inline README.md]: Typo, client's needs

File: docs/component.md [hunk 2]
```diff
@@ -3,353 +3,581 @@ id: component
...
 ---
 
-## The concept of a component
 We trust in the web.
 
-That's why we decided to write biotope and the biotope element with mostly vanilla web technologies
-and polyfill it untill the right time is there.
-On of these fancy new technologies is web components and all the proposals belonging to it (shadow
-dom, custom elements, scoped styling).
+`biotope-element` is an extension of a vanilla a HTMLElement with some nice features on top. So
```
Comment (pmpinto) [inline docs/component.md]: > of a vanilla a HTMLElement

Should be:
> of a vanilla HTMLElement

File: docs/component.md [hunk 3]
```diff
@@ -3,353 +3,581 @@ id: component
...
+MyButton.register();
+```
+
+If you implement the example above, the `render` function will only be set during the constructor,
+and not when calling the `register` function. This will also cause your component to render nothing.
+
+### Partials
+The `html` function can be used to convert string HTML to actual HTML components. In this sense, if
+you need to render a non-HTML string, then you can simply interpolate it inside the template string
+that you pass to the `this.html` function. You can even use regular Javascript - inline `if`s are
```
Comment (pmpinto) [inline docs/component.md]: > Javascript

Should be:

> JavaScript

Comment (pmpinto) [reply docs/component.md]: > inline `if`s

Could be:

> inline conditions

Or something similar, since you're not actually using `if` and it's easier to read.

File: docs/component.md [hunk 4]
```diff
@@ -3,353 +3,581 @@ id: component
...
+  <div># All done!</div>
+</my-button>
 ```
 
+## Attributes
+This is the main way of passing information to any component.
+
+Every component has its own attributes. If you have any experience with adding for example
+`class`es, `id`s, `src`s or `data-*` attributes on regular HTMLElements, then you already know how
+to add/modify/remove attributes on `biotope-element`s.
```
Comment (pmpinto) [inline docs/component.md]: > on `biotope-element`s.

Could be:

> in `biotope-element`.

File: docs/component.md [hunk 5]
```diff
@@ -3,353 +3,581 @@ id: component
...
+
+To add or modify an attribute to a `biotope-element` component, you can simply do:
+
 ```html
 <my-button></my-button>
+<script>
+  const myElement = document.getElementsByTagName('my-button');
```
Comment (pmpinto) [inline docs/component.md]: `getElementsByTagName` returns an `HTMLCollection`, so `setAttribute()` can't be used directly.

Consider:
`document.querySelector('my-button')`

File: docs/component.md [hunk 6]
```diff
@@ -3,353 +3,581 @@ id: component
...
-<my-button>🎰 bar</my-button>
+<my-button example-attribute="Hello Foo World"></my-button>
+<script>
+  const myElement = document.getElementsByTagName('my-button');
+  myElement.setAttribute('example-attribute', 'Hello Foo World');
+</script>
 ```
 
-In typescript, it would look like this:
+Additionally, you can use a shortened notation that we provide in `biotope-element`s.
```
Comment (pmpinto) [inline docs/component.md]: > shortened notation

Is usually known as:

> short notation

Comment (pmpinto) [reply docs/component.md]: >  in `biotope-element`s.

Could be:

>  in `biotope-element`.

File: docs/component.md [hunk 7]
```diff
@@ -3,353 +3,581 @@ id: component
...
+### Props
+Since attributes are the main API for passing information to your component, we decided to add props
+to our components to simplify and streamline this process. This means that, props are the result of
+picking and parsing of all the attributes of the component. However, this is only efficient if the
+component knows which attributes to watch and which ones to ignore.
 
-  protected readonly defaultProps: MyButtonProps = {
-    foo: 'bar',
-  };
+A prop change will always trigger a new render. Consider this next example component, with a "greet"
```
Comment (pmpinto) [inline docs/component.md]: > "greet"

Should be:

> `greet`

File: docs/styling.md [hunk 8]
```diff
@@ -3,126 +3,193 @@ id: styling
 title: Styling
 ---
 
-## Adding styles
-As every component uses Shadow DOM by default, you have to put the styles inside the shadow root.
-You can either import your css from an external file (recommended way for bio-element). In this case
-your imported style has to be a string with the css inside:
+Web components use a shadow DOM to write their HTML. The shadow DOM has some interesting features
+that make it very different from writing to an element using `innerHTML`. Styling-wise, the shadow
+DOM scopes the style, so that styles from the page does not leak into the component and style and
```
Comment (pmpinto) [inline docs/styling.md]: > styles from the page does not leak

Should be:

> styles from the page do not leak

File: docs/styling.md [hunk 9]
```diff
@@ -3,126 +3,193 @@ id: styling
...
+`div`, like we've shown you before. Here's a simple example:
+
+```javascript
+// render function
+return this.html`
+  <div>
+    Hello World 🐤
+  </div>
+  <style>
+    color: red;
```
Comment (pmpinto) [inline docs/styling.md]: CSS selector is missing.

Same for line 52.

File: docs/styling.md [hunk 10]
```diff
@@ -3,126 +3,193 @@ id: styling
...
+  </style>
+`;
 ```
 
-or use inline styles:
+### createStyle function
+We provide you with a pretty cool function to do most of this automatically - the `createStyle`
+function. It can take either take a string (or an object with a `toString` function) and convert it
```
Comment (pmpinto) [inline docs/styling.md]: > It can take either take a string

Should have been:

> It can either take a string

But... I was expecting an `or` and didn't find it. Like:

> It can either be A or B.

File: docs/lifecycle.md [hunk 11]
```diff
@@ -0,0 +1,203 @@
...
+When the component is removed from the page, only the `disconnectedCallback` hooks gets triggered.
+
+Below we provide a simple explanation and simple examples of how all of these hooks work, when they
+are called and what you should be doing on each one. This however does not replace any Web Component
+hook documentation.
+
+### constructor
+Simple enough to understand - just a class constructor.
+
+This is the point of entry point for the creation of the instance of your component. Here are some
```
Comment (pmpinto) [inline docs/lifecycle.md]: > the point of entry point

Should be:

> the entry point

File: docs/lifecycle.md [hunk 12]
```diff
@@ -0,0 +1,203 @@
...
+examples of things you can/should do in this function:
+  - bind functions to the `this` context;
```
Comment (pmpinto) [inline docs/lifecycle.md]: Either use a trailing `;` or don't.

File: docs/lifecycle.md [hunk 13]
```diff
@@ -0,0 +1,203 @@
...
+This is where you can typically:
+  - query the DOM (upwards or downwards)
+  - add event listeners to the component itself
+  - read/write to your slot (`innerHTML`) - if need be…
+
+> __⚠️ Important:__ In vanilla Web Components, this hook is called only **after** `attributeChangedCallback`.
+In `biotope-element`however, this order is switched for render optimization purposes.
+
+### attributeChangedCallback
+This hooks is called when an attribute has changed on the component. It can trigger a new render if
```
Comment (pmpinto) [inline docs/lifecycle.md]: > This hooks

Should be:

> This hook

File: docs/lifecycle.md [hunk 14]
```diff
@@ -0,0 +1,203 @@
...
+This will only result in one `render` being called though.
+
+### rendered
+This is the second hook provided by `biotope-element`. This hook will only be called when a `render`
+has finished and the DOM has been updated. If two `render`s are called, this hook will also be
+called twice.
+
+This means that, in this hook, you can:
+  - query the newly created element for anything you want
+  - add some attributes that are can only be added after a DOM element is rendered (like the
```
Comment (pmpinto) [inline docs/lifecycle.md]: > that are can only be added

Should be:

> that can only be added

File: docs/lifecycle.md [hunk 15]
```diff
@@ -0,0 +1,203 @@
...
+  connectedCallback() {
+    console.log('second');
+  }
+
+  attributeChangedCallback(name, previous, current) {
+    super.attributeChangedCallback(name, previous, current);
+    console.log('third', name);
+  }
+
+  render(): HTMLFragment {
```
Comment (pmpinto) [inline docs/lifecycle.md]: This is not TypeScript, does `HTMLFragment` here still make sense?

File: docs/attr-parsing.md [hunk 16]
```diff
@@ -0,0 +1,215 @@
+---
+id: attr-parsing
+title: Attribute Parsing
+---
+
+In regular HTMLElements, you can set attributes like `class`, `id`, `title`, `lang`, … These
+attributes are all strings, as you might discover when trying to set one of them to a Javascript
```
Comment (pmpinto) [inline docs/attr-parsing.md]: > Javascript

Should be:

> JavaScript

File: docs/attr-parsing.md [hunk 17]
```diff
@@ -0,0 +1,215 @@
...
+object but you ending up with something like `class="[object Object]"`.
```
Comment (pmpinto) [inline docs/attr-parsing.md]: > but you ending up with something

Should be:

> and you will end up with something

File: docs/attr-parsing.md [hunk 18]
```diff
@@ -0,0 +1,215 @@
...
+instead of a string, you can simply do:
+
+```javascript
+MyButton.attributes = [
+  'text',
+  { name: 'disabled', type: 'boolean' },
+];
+```
+
+This will automatically try to force the value it receives into a javascript boolean type - it does
```
Comment (pmpinto) [inline docs/attr-parsing.md]: > javascript

Should be:

> JavaScript

Comment (pmpinto) [reply docs/attr-parsing.md]: > boolean

Should be:

> `boolean`

File: docs/attr-parsing.md [hunk 19]
```diff
@@ -0,0 +1,215 @@
...
+    text=${'Hello World'}
+    another-text="Hello World too"
+    example-array=${myArray}
+    another-attribute=${JSON.stringify(myArray)}
+  ></my-button>
+`;
+```
+
+If we now consider that the child treats `text` and `another-text` both as strings `example-array`
+and `another-array` both as of type "array", then we can safely deduce that the attribute `another-array`
```
Comment (pmpinto) [inline docs/attr-parsing.md]: > "array"

Should be:

> `array`

File: docs/attr-parsing.md [hunk 20]
```diff
@@ -0,0 +1,215 @@
...
+will definitely take longer to parse due to being converted to/from string into an array.
```
Comment (pmpinto) [inline docs/attr-parsing.md]: > string into an array.

Should be:

> `string` into an `array`.

File: docs/attr-parsing.md [hunk 21]
```diff
@@ -0,0 +1,215 @@
...
+<my-form>
+  <my-button
+    text="Hello World"
+    another-text="Hello World too"
+    another-attribute="[1,2,3,4]"
+  ></my-button>
+</my-form>
+```
+
+This is especially important when passing functions to a child element since Javascript functions do
```
Comment (pmpinto) [inline docs/attr-parsing.md]: > Javascript

Should be:

> JavaScript

File: docs/attr-parsing.md [hunk 22]
```diff
@@ -0,0 +1,215 @@
...
+not retain their context when being converter to/from a string.
```
Comment (pmpinto) [inline docs/attr-parsing.md]: > when being converter to/from a string.

Should be:

> when being converted to/from a `string`.

File: docs/typescript.md [hunk 23]
```diff
@@ -0,0 +1,120 @@
+---
+id: typescript
+title: Typescript
```
Comment (pmpinto) [inline docs/typescript.md]: > Typescript

Should be:

> TypeScript

Comment (pmpinto) [reply docs/typescript.md]: Same for L6.

Comment (pmpinto) [reply docs/typescript.md]: And everywhere else throughout this file.

File: docs/typescript.md [hunk 24]
```diff
@@ -0,0 +1,120 @@
...
+
+MyButton.register();
+```
+
+This is the parent component:
+
+```typescript
+// my-form.ts
+import Component, { createRef, HTMLFragment } from '@biotope/element';
+import { MyButton, MuButtonEventPressed } from './my-button';
```
Comment (pmpinto) [inline docs/typescript.md]: > `MuButtonEventPressed`

Should be:

> `MyButtonEventPressed`

File: docs/typescript.md [hunk 25]
```diff
@@ -0,0 +1,120 @@
...
+  public static componentName = 'my-form';
+
+  public static dependencies = [
+    MyButton as typeof Component,
+  ];
+
+  private refs = {
+    myButton: createRef<MyButton>(),
+    input: createRefCallback<HTMLInputElement>(
+      () => this.shadowRoot.querySelector('my-fancy-class'),
```
Comment (pmpinto) [inline docs/typescript.md]: > `my-fancy-class`

Should be:

> `.my-fancy-class`

File: docs/peers.md [hunk 26]
```diff
@@ -0,0 +1,40 @@
...
+
+## biotope
+`biotope` is our CLI tool to create projects with Biotope. This also includes generating `biotope-element`
+components with:
+
+```bash
+biotope generate
+```
+
+You can install it with `npm install biotope` and can read more about it in our [github repo](https://github.com/biotope/biotope).
```
Comment (pmpinto) [inline docs/peers.md]: > github

Should be:

> GitHub

File: docs/peers.md [hunk 27]
```diff
@@ -0,0 +1,40 @@
...
+like so:
+
+```html
+<my-button data-resources="[{paths: ['path/to/js/my-button.js']}]"></my-button>
+```
+
+This will load the script and attach it to the dom.
+
+You can install it with `npm install @biotope/resource-loader` and you can read more about it in our
+[github repo](https://github.com/biotope/biotope-resource-loader).
```
Comment (pmpinto) [inline docs/peers.md]: > github

Should be:

> GitHub

File: docs/peers.md [hunk 28]
```diff
@@ -0,0 +1,40 @@
...
+## @biotope/quality-gate
+`@biotope/quality-gate` makes it easy for you to lint the components you built with `biotope-element`
+by running a simple command:
+
+```bash
+> biotope-quality-gate
+```
+
+We advise this package to be installed as a dev dependency. You can install it with `npm install @biotope/quality-gate -D`
+and you can read more about it in our [github repo](https://github.com/biotope/biotope-quality-gate).
```
Comment (pmpinto) [inline docs/peers.md]: > github

Should be:

> GitHub

File: docs/integration.md [hunk 29]
```diff
@@ -0,0 +1,26 @@
...
+
+Currently there are no dev tools or extensions for browsers and IDEs that are supported by us.
+
+However, we can show you how we work efficiently with `biotope-element` on these editors.
+
+## VSCode
+
+Use `@biotope/quality-gate` for code checking. You can find out more about how to easily set it up
+on VSCode in [our repo](https://github.com/biotope/biotope-quality-gate). This will cover most of
+your use-cases and prevent you from writing no-so-good code.
```
Comment (pmpinto) [inline docs/integration.md]: > no-so-good

Should be:

> not-so-good

File: docs/migration.md [hunk 30]
```diff
@@ -3,156 +3,9 @@ id: migration
...
-`<demo-component data-resources="[{paths: ['components/demo-component/index.js']}]"></demo-component>`
-
-
-# Migrating from v3 to v4
-
-## `wire` function removal and `html` function update
+### `wire` function removal and `html` function update
 
 TLDR: `wire` function is gone - you can use `html` for everything now. The `render` function now
 needs it's return value to be the result of `this.html`, otherwise nothing will get rendered.
```
Comment (pmpinto) [inline docs/migration.md]: > needs it's return value to be the result of `this.html`

Could be:

> needs to return the result of `this.html`

File: docs/migration.md [hunk 31]
```diff
@@ -313,7 +166,7 @@ class MyComponent extends Component {
 }
 ```
 
-## Automatic types
+### Automatic types
 
 TLDR: no more converting strings to js types - just add a `type` property to each attribute (manual
```
Comment (pmpinto) [inline docs/migration.md]: > js types

Should be:

> JavaScript types
